CI: graph-lifetime tests flaky on free-threaded 3.14t jobs since #2371

## Describe the bug

Since #2371 merged (2026-07-15 16:55 UTC), CI is flaky on **free-threaded Python 3.14t** test jobs: four tests in [`cuda_core/tests/graph/test_graph_definition_lifetime.py`](https://github.com/NVIDIA/cuda-python/blob/58f87330386f672cf29c716ca84ba8e832967300/cuda_core/tests/graph/test_graph_definition_lifetime.py) intermittently fail with

```
AssertionError: condition not satisfied within 5.0s
```

raised by the [`_wait_until`](https://github.com/NVIDIA/cuda-python/blob/58f87330386f672cf29c716ca84ba8e832967300/cuda_core/tests/graph/test_graph_definition_lifetime.py#L52-L75) polling helper ([`_FINALIZE_TIMEOUT = 5.0`](https://github.com/NVIDIA/cuda-python/blob/58f87330386f672cf29c716ca84ba8e832967300/cuda_core/tests/graph/test_graph_definition_lifetime.py#L29)):

- `test_kernel_args_buffer_lifetime` (added in #2041)
- `test_memcpy_buffer_allocations_released_after_graph_destroyed` (#2280)
- `test_slot_owners_released_after_graph_destroyed` (#2280)
- `test_pending_call_queue_saturation_preserves_cleanup` (#2371)

## Impact (2026-07-15 → 2026-07-21)

Survey of all Actions runs 2026-06-15 → 2026-07-22 (246 failed/re-run runs classified by job-log signature):

- **16 of 77 CI runs (~21%) hit this signature**, including 5 runs on `main` (push + scheduled). 11 were on PRs touching nothing near graph code (e.g. #2374, #2402).
- 7 of the 16 were re-run to green — one needed [6 attempts](https://github.com/NVIDIA/cuda-python/actions/runs/29602432384) — so red-run counts understate it.
- **Zero occurrences** in ~300 completed runs 2026-06-15 → 2026-07-14. #2280 (merged 2026-07-07) ran clean for 8 days; the first hit is 4.5 h after #2371 merged (2026-07-15 21:26 UTC, [run 29451966527](https://github.com/NVIDIA/cuda-python/actions/runs/29451966527)).
- Only 3.14t jobs are affected — across win-64 / linux-64 / linux-aarch64, CUDA 12.9.1 / 13.0.2 / 13.3.0, local & wheels, T4 / L4 / A100. Never observed on regular-GIL jobs.

## Failing jobs (all 18 instances)

| Failing CI pipeline | CI log |
|---|---|
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-15 · PR 2374](https://github.com/NVIDIA/cuda-python/actions/runs/29451966527/job/87480738831) |
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-16 · PR 2242](https://github.com/NVIDIA/cuda-python/actions/runs/29499296934/job/87629352733) |
| Test linux-aarch64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 | [2026-07-17 · PR 2379](https://github.com/NVIDIA/cuda-python/actions/runs/29584165382/job/87899244042) |
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-17 · main](https://github.com/NVIDIA/cuda-python/actions/runs/29555550182/job/87809528510) |
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-17 · main](https://github.com/NVIDIA/cuda-python/actions/runs/29602432384/job/87962162694) |
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-19 · main](https://github.com/NVIDIA/cuda-python/actions/runs/29666641194/job/88139923507) |
| Test linux-64 / Python 3.14t, CUDA 12.9.1 (local), GPU t4 | [2026-07-20 · PR 2390](https://github.com/NVIDIA/cuda-python/actions/runs/29724266375/job/88295548106) |
| Test linux-64 / Python 3.14t, CUDA 13.0.2 (local), GPU l4 | [2026-07-20 · PR 2242](https://github.com/NVIDIA/cuda-python/actions/runs/29768699624/job/88444074974) |
| Test linux-aarch64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 | [2026-07-20 · main](https://github.com/NVIDIA/cuda-python/actions/runs/29719293380/job/88279928623) |
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-20 · PR 2242](https://github.com/NVIDIA/cuda-python/actions/runs/29768699624/job/88446543396) |
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-20 · PR 2394](https://github.com/NVIDIA/cuda-python/actions/runs/29777373005/job/88475693911) |
| Test linux-64 / Python 3.14t, CUDA 12.9.1 (local), GPU t4 | [2026-07-21 · PR 2398](https://github.com/NVIDIA/cuda-python/actions/runs/29849525172/job/88701835631) |
| Test linux-64 / Python 3.14t, CUDA 12.9.1 (local), GPU t4 | [2026-07-21 · PR 2402](https://github.com/NVIDIA/cuda-python/actions/runs/29876578909/job/88790610219) |
| Test linux-aarch64 / Python 3.14t, CUDA 13.3.0 (local), GPU l4 | [2026-07-21 · main](https://github.com/NVIDIA/cuda-python/actions/runs/29846402084/job/88690759075) |
| Test linux-aarch64 / Python 3.14t, CUDA 13.3.0 (local), GPU l4 | [2026-07-21 · PR 2383](https://github.com/NVIDIA/cuda-python/actions/runs/29855714919/job/88722171657) |
| Test win-64 / Python 3.14t, CUDA 12.9.1 (local), GPU l4 (TCC) | [2026-07-21 · PR 2390](https://github.com/NVIDIA/cuda-python/actions/runs/29796301934/job/88530809561) |
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-21 · PR 2392](https://github.com/NVIDIA/cuda-python/actions/runs/29825220814/job/88621734327) |
| Test win-64 / Python 3.14t, CUDA 13.0.2 (wheels), GPU a100 (MCDM) | [2026-07-21 · PR 2398](https://github.com/NVIDIA/cuda-python/actions/runs/29849525172/job/88704708482) |

(Excluded from the count: graph PRs failing their own branches — #2280, #2333, #2357 — and the 2026-07-17 gb300 nightly, which failed machine-wide with `ValueError: Cannot set NULL context as current` and looks unrelated.)

## How to Reproduce

No deterministic local repro; it fires in roughly 1 in 5 CI runs on any 3.14t GPU test job since 2026-07-15. The four tests above are the ones to stress.

## Expected behavior

Graph-lifetime cleanup assertions pass reliably on free-threaded builds.

## Likely mechanism

#2371 defers user-object payload cleanup through a pending-call queue. On free-threaded Python (deferred reference counting), on a loaded runner, cleanup can lag past the 5 s `_FINALIZE_TIMEOUT` budget, so assertions observing release (weakref death, allocation counters, queue drain) time out. Three of the four tests predate #2371 and only began flaking once it landed.

Possible directions:

- Raise `_FINALIZE_TIMEOUT` on free-threaded builds (it is already 30 s under compute-sanitizer).
- Make `_wait_until` actively drive the pending-call queue instead of only `gc.collect()` + sleep.
- Bound cleanup latency in the implementation (e.g. drain the queue opportunistically on API calls).

-- Leo's bot

--- a/cuda_core/tests/graph/test_graph_definition_lifetime.py
+++ b/cuda_core/tests/graph/test_graph_definition_lifetime.py
@@ -386,7 +386,7 @@ def test_user_object_cleanup_is_coalesced_on_python_thread(init_cuda):
 @pytest.mark.agent_authored(model="gpt-5.6")
 def test_pending_call_queue_saturation_preserves_cleanup(tmp_path):
     """A full CPython queue neither strands nor mis-threads cleanup."""
-    code = textwrap.dedent(
+    code = f"timeout = {_FINALIZE_TIMEOUT!r}\n" + textwrap.dedent(
         """
         import ctypes
         import gc
@@ -395,9 +395,6 @@ def test_pending_call_queue_saturation_preserves_cleanup(tmp_path):
 
         from cuda.core import Device
         from cuda.core.graph import GraphDefinition
-        from cuda_python_test_helpers import under_compute_sanitizer
-
-        timeout = 30.0 if under_compute_sanitizer() else 5.0
 
         class ThreadRecordingCallback:
             def __init__(self, finalized_threads):

--- a/cuda_core/tests/graph/test_graph_definition_lifetime.py
+++ b/cuda_core/tests/graph/test_graph_definition_lifetime.py
@@ -384,51 +384,102 @@ def test_user_object_cleanup_is_coalesced_on_python_thread(init_cuda):
 
 
 @pytest.mark.agent_authored(model="gpt-5.6")
-def test_pending_call_queue_saturation_preserves_cleanup(init_cuda):
+def test_pending_call_queue_saturation_preserves_cleanup(tmp_path):
     """A full CPython queue neither strands nor mis-threads cleanup."""
-    pending_callback_type = ctypes.CFUNCTYPE(ctypes.c_int, ctypes.c_void_p)
-    add_pending_call = ctypes.pythonapi.Py_AddPendingCall
-    add_pending_call.argtypes = [pending_callback_type, ctypes.c_void_p]
-    add_pending_call.restype = ctypes.c_int
+    code = textwrap.dedent(
+        """
+        import ctypes
+        import gc
+        import threading
+        import time
 
-    @pending_callback_type
-    def noop_pending_call(_):
-        return 0
+        from cuda.core import Device
+        from cuda.core.graph import GraphDefinition
+        from cuda_python_test_helpers import under_compute_sanitizer
 
-    finalized_threads = []
-    main_thread = threading.get_ident()
-    first_callback = _ThreadRecordingCallback(finalized_threads)
-    first_graph = GraphDefinition()
-    first_graph.callback(first_callback)
-    graph_holder = [first_graph]
-    worker_done = threading.Event()
-    queue_was_full = []
+        timeout = 30.0 if under_compute_sanitizer() else 5.0
 
-    del first_callback, first_graph
+        class ThreadRecordingCallback:
+            def __init__(self, finalized_threads):
+                self.finalized_threads = finalized_threads
 
-    def fill_queue_and_destroy():
-        while add_pending_call(noop_pending_call, None) == 0:
-            pass
-        queue_was_full.append(True)
-        graph_holder.clear()
-        worker_done.set()
+            def __call__(self):
+                pass
 
-    worker = threading.Thread(target=fill_queue_and_destroy)
-    worker.start()
-    assert worker_done.wait(timeout=5)
-    worker.join()
-    assert queue_was_full == [True]
+            def __del__(self):
+                self.finalized_threads.append(threading.get_ident())
 
-    # Preparing another attachment retries the first payload if its scheduling
-    # attempt observed the full queue. CUDA may also invoke its destructor
-    # later, after pending-call queue space is available.
-    second_callback = _ThreadRecordingCallback(finalized_threads)
-    second_graph = GraphDefinition()
-    second_graph.callback(second_callback)
-    del second_callback, second_graph
+        def wait_until(predicate):
+            deadline = time.monotonic() + timeout
+            while True:
+                gc.collect()
+                if predicate():
+                    return
+                if time.monotonic() >= deadline:
+                    break
+                time.sleep(0)
+                time.sleep(0.02)
+            raise AssertionError(f"condition not satisfied within {timeout}s")
 
-    _wait_until(lambda: len(finalized_threads) == 2)
-    assert set(finalized_threads) == {main_thread}
+        Device(0).set_current()
+
+        pending_callback_type = ctypes.CFUNCTYPE(ctypes.c_int, ctypes.c_void_p)
+        add_pending_call = ctypes.pythonapi.Py_AddPendingCall
+        add_pending_call.argtypes = [pending_callback_type, ctypes.c_void_p]
+        add_pending_call.restype = ctypes.c_int
+        make_pending_calls = ctypes.pythonapi.Py_MakePendingCalls
+        make_pending_calls.argtypes = []
+        make_pending_calls.restype = ctypes.c_int
+
+        @pending_callback_type
+        def noop_pending_call(_):
+            return 0
+
+        finalized_threads = []
+        main_thread = threading.get_ident()
+        first_callback = ThreadRecordingCallback(finalized_threads)
+        first_graph = GraphDefinition()
+        first_graph.callback(first_callback)
+        graph_holder = [first_graph]
+        worker_done = threading.Event()
+        queue_was_full = []
+
+        del first_callback, first_graph
+
+        def fill_queue_and_destroy():
+            while add_pending_call(noop_pending_call, None) == 0:
+                pass
+            queue_was_full.append(True)
+            graph_holder.clear()
+            worker_done.set()
+
+        worker = threading.Thread(target=fill_queue_and_destroy)
+        worker.start()
+        assert worker_done.wait(timeout=5)
+        worker.join()
+        assert queue_was_full == [True]
+
+        # Free space before the cuda-core entry that retries cleanup scheduling.
+        assert make_pending_calls() == 0
+
+        second_callback = ThreadRecordingCallback(finalized_threads)
+        second_graph = GraphDefinition()
+        second_graph.callback(second_callback)
+        del second_callback, second_graph
+
+        wait_until(lambda: len(finalized_threads) == 2)
+        assert set(finalized_threads) == {main_thread}
+        """
+    )
+    result = subprocess.run(  # noqa: S603 - controlled interpreter probe
+        [sys.executable, "-c", code],
+        capture_output=True,
+        text=True,
+        timeout=60,
+        # Isolate the process-global pending-call queue from parallel tests.
+        cwd=tmp_path,
+    )
+    assert result.returncode == 0, result.stderr
 
 
 @pytest.mark.agent_authored(model="gpt-5.6")

